src/selectors/prevThought.ts

```diff
@@ -0,0 +1,124 @@
+import Path from '../@types/Path'
+import State from '../@types/State'
+import getChildren, { isVisible } from '../selectors/getChildren'
+import isContextViewActive from '../selectors/isContextViewActive'
+import prevContext from '../selectors/prevContext'
+import prevSibling from '../selectors/prevSibling'
+import rootedParentOf from '../selectors/rootedParentOf'
+import appendToPath from '../util/appendToPath'
+import hashPath from '../util/hashPath'
+import head from '../util/head'
+import isRoot from '../util/isRoot'
+import parentOf from '../util/parentOf'
+import getContextsSortedAndRanked from './getContextsSortedAndRanked'
+import getThoughtById from './getThoughtById'
+import simplifyPath from './simplifyPath'
+
+/** Gets the last visible descendant of a thought. */
+const lastVisibleDescendant = (state: State, path: Path): Path => {
+  // Use simplifyPath only if we're in context view mode
+  const isInContextView = Object.keys(state.contextViews).length > 0 && path.length > 1
+  const effectivePath = isInContextView ? simplifyPath(state, path) : path
+  const id = head(effectivePath)
+
+  // Only traverse children if the thought or its parent is expanded
+  if (!state.expanded[hashPath(effectivePath)]) return path
+
+  const children = getChildren(state, id)
+  // If no children, return current path
+  if (children.length === 0) return path
+
+  const lastChild = children[children.length - 1]
+  return lastVisibleDescendant(state, appendToPath(path, lastChild.id))
+}
+
+/** Gets the last context in a context view. */
+const lastContext = (state: State, path: Path): Path | null => {
+  const thought = getThoughtById(state, head(path))
+  const contexts = getContextsSortedAndRanked(state, thought.value).filter(
+    context => state.showHiddenThoughts || isVisible(state, getThoughtById(state, context.id)),
+  )
+
+  // if context view is empty or no visible contexts, return null
+  if (contexts.length === 0) return null
+
+  // get the last visible context
+  const lastContext = getThoughtById(state, contexts[contexts.length - 1]?.id)
+  if (!state.expanded[hashPath(appendToPath(path, lastContext.parentId))])
+    return appendToPath(path, lastContext.parentId)
+  return lastVisibleDescendant(state, appendToPath(path, lastContext.parentId))
+}
+
+/**
+ * Gets the last visible descendant of the previous sibling.
+ * If the previous sibling is hidden, recursively tries the next previous sibling.
+ */
+const getLastVisibleDescendantOfPreviousSibling = (
```

Comment:
It would probably be cleaner to separate this function into two:

1. Get the previous visible sibling
2. Get the last visible descendant of the previous sibling